[wasi] Single file build fails with AOT

This will show up after adding new AOT tests on wasi in https://github.com/dotnet/runtime/pull/95281.

```
        Stack Trace:
          C:\Users\source\repos\runtime-fork\src\mono\wasi\Wasi.Build.Tests\BuildTestBase.cs(445,0): at Wasm.Build.Tests.BuildTestBase.AssertBuild(String args, String label, Boolean expectSuccess, IDictionary`2 envVars, Nullable`1 timeoutMs)
          C:\Users\source\repos\runtime-fork\src\mono\wasi\Wasi.Build.Tests\BuildTestBase.cs(255,0): at Wasm.Build.Tests.BuildTestBase.BuildProject(BuildArgs buildArgs, String id, BuildProjectOptions options)
          C:\Users\source\repos\runtime-fork\src\mono\wasi\Wasi.Build.Tests\BuildThenPublish.cs(50,0): at Wasi.Build.Tests.BuildThenPublishTests.ConsoleBuildThenPublish(String config, Boolean aot, Boolean singleFileBundle)
             at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
             at System.Reflection.MethodBaseInvoker.InvokeDirectByRefWithFewArgs(Object obj, Span`1 copyOfArgs, BindingFlags invokeAttr)
        Output:
          [] Executing (Captured Output) - C:\Users\source\repos\runtime-fork\artifacts\bin\dotnet-latest\dotnet.exe new wasiconsole  -  in pwd C:\Users\source\repos\runtime-fork\artifacts\bin\Wasi.Build.Tests\Release\net8.0\wasi-wasm\wbt\Debug_dslxjqc1_otc
          [] The template "Wasi Console App" was created successfully.
          [] 
          -------- Building ---------
          Binlog path: C:\Users\source\repos\runtime-fork\artifacts\bin\Wasi.Build.Tests\Release\net8.0\wasi-wasm\xharness-output\logs\Debug_dslxjqc1_otc\Debug_dslxjqc1_otc.binlog
          Building Debug_dslxjqc1_otc in C:\Users\source\repos\runtime-fork\artifacts\bin\Wasi.Build.Tests\Release\net8.0\wasi-wasm\wbt\Debug_dslxjqc1_otc
          Running C:\Users\source\repos\runtime-fork\artifacts\bin\dotnet-latest\dotnet.exe build  /p:Configuration=Debug /bl:"C:\Users\source\repos\runtime-fork\artifacts\bin\Wasi.Build.Tests\Release\net8.0\wasi-wasm\xharness-output\logs\Debug_dslxjqc1_otc\Debug_dslxjqc1_otc.binlog" /nologo /v:minimal
          WorkingDirectory: C:\Users\source\repos\runtime-fork\artifacts\bin\Wasi.Build.Tests\Release\net8.0\wasi-wasm\wbt\Debug_dslxjqc1_otc
          Setting environment variables for execution:
          	DOTNET_ROOT = C:\Users\source\repos\runtime-fork\artifacts\bin\dotnet-latest
          	DOTNET_INSTALL_DIR = C:\Users\source\repos\runtime-fork\artifacts\bin\dotnet-latest
          	DOTNET_MULTILEVEL_LOOKUP = 0
          	DOTNET_SKIP_FIRST_TIME_EXPERIENCE = 1
          	PATH = C:\Users\source\repos\runtime-fork\artifacts\bin\dotnet-latest;C:\Users\source\repos\runtime-fork\artifacts\bin\\dotnet-latest;C:\Users\source\repos\runtime-fork\artifacts\obj\wasmtime\
          	EM_WORKAROUND_PYTHON_BUG_34780 = 1
          	WASI_SDK_PATH = C:\Users\source\repos\runtime-fork\src\mono\wasi\wasi-sdk
          	NUGET_PACKAGES = C:\Users\source\repos\runtime-fork\artifacts\bin\Wasi.Build.Tests\Release\net8.0\wasi-wasm\wbt\nuget\Debug_dslxjqc1_otc
          [Debug_dslxjqc1_otc]   Determining projects to restore...
          [Debug_dslxjqc1_otc]   Restored C:\Users\source\repos\runtime-fork\artifacts\bin\Wasi.Build.Tests\Release\net8.0\wasi-wasm\wbt\Debug_dslxjqc1_otc\Debug_dslxjqc1_otc.csproj (in 3,58 sec).
          [Debug_dslxjqc1_otc] C:\Users\source\repos\runtime-fork\artifacts\bin\dotnet-latest\sdk\9.0.100-alpha.1.23564.26\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(311,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy [C:\Users\source\repos\runtime-fork\artifacts\bin\Wasi.Build.Tests\Release\net8.0\wasi-wasm\wbt\Debug_dslxjqc1_otc\Debug_dslxjqc1_otc.csproj]
          [Debug_dslxjqc1_otc]   Debug_dslxjqc1_otc -> C:\Users\source\repos\runtime-fork\artifacts\bin\Wasi.Build.Tests\Release\net8.0\wasi-wasm\wbt\Debug_dslxjqc1_otc\bin\Debug\net8.0\wasi-wasm\Debug_dslxjqc1_otc.dll
          [Debug_dslxjqc1_otc]   ** MicrosoftNetCoreAppRuntimePackDir : 'C:\Users\source\repos\runtime-fork\artifacts\bin\dotnet-latest\packs\Microsoft.NETCore.App.Runtime.Mono.wasi-wasm\9.0.0-dev'
          [Debug_dslxjqc1_otc]   MicrosoftNetCoreAppRuntimePackRidNativeDir: C:\Users\source\repos\runtime-fork\artifacts\bin\dotnet-latest\packs\Microsoft.NETCore.App.Runtime.Mono.wasi-wasm\9.0.0-dev\runtimes\wasi-wasm\native\
          [Debug_dslxjqc1_otc]   Compiling native assets using C:\Users\source\repos\runtime-fork\src\mono\wasi\wasi-sdk\bin\clang.exe with -O1. This may take a while ...
          [Debug_dslxjqc1_otc]   [1/5] stubs.c -> stubs.o [took 0,26s]
          [Debug_dslxjqc1_otc]   [2/5] synthetic-pthread.c -> synthetic-pthread.o [took 0,30s]
          [Debug_dslxjqc1_otc]   [3/5] pinvoke.c -> pinvoke.o [took 0,32s]
          [Debug_dslxjqc1_otc]   [4/5] main.c -> main.o [took 0,35s]
          [Debug_dslxjqc1_otc] C:\Users\source\repos\runtime-fork\artifacts\bin\dotnet-latest\packs\Microsoft.NET.Runtime.WebAssembly.Wasi.Sdk\9.0.0-dev\Sdk\WasiApp.Native.targets(273,5): error : Failed to compile C:\Users\source\repos\runtime-fork\artifacts\bin\dotnet-latest\packs\Microsoft.NETCore.App.Runtime.Mono.wasi-wasm\9.0.0-dev\runtimes\wasi-wasm\native\src\driver.c -> C:\Users\source\repos\runtime-fork\artifacts\bin\Wasi.Build.Tests\Release\net8.0\wasi-wasm\wbt\Debug_dslxjqc1_otc\obj\Debug\net8.0\wasi-wasm\wasm\for-build\driver.o [C:\Users\source\repos\runtime-fork\artifacts\bin\Wasi.Build.Tests\Release\net8.0\wasi-wasm\wbt\Debug_dslxjqc1_otc\Debug_dslxjqc1_otc.csproj]
          [Debug_dslxjqc1_otc] C:\Users\source\repos\runtime-fork\artifacts\bin\dotnet-latest\packs\Microsoft.NET.Runtime.WebAssembly.Wasi.Sdk\9.0.0-dev\Sdk\WasiApp.Native.targets(273,5): error : C:\Users\source\repos\runtime-fork\artifacts\bin\Wasi.Build.Tests\Release\net8.0\wasi-wasm\wbt\Debug_dslxjqc1_otc>C:\WINDOWS\System32\chcp.com 65001 1>nul  [C:\Users\source\repos\runtime-fork\artifacts\bin\Wasi.Build.Tests\Release\net8.0\wasi-wasm\wbt\Debug_dslxjqc1_otc\Debug_dslxjqc1_otc.csproj]
          [Debug_dslxjqc1_otc] C:\Users\source\repos\runtime-fork\artifacts\bin\dotnet-latest\packs\Microsoft.NET.Runtime.WebAssembly.Wasi.Sdk\9.0.0-dev\Sdk\WasiApp.Native.targets(273,5): error : C:\Users\source\repos\runtime-fork\artifacts\bin\Wasi.Build.Tests\Release\net8.0\wasi-wasm\wbt\Debug_dslxjqc1_otc>setlocal [C:\Users\source\repos\runtime-fork\artifacts\bin\Wasi.Build.Tests\Release\net8.0\wasi-wasm\wbt\Debug_dslxjqc1_otc\Debug_dslxjqc1_otc.csproj]
          [Debug_dslxjqc1_otc] C:\Users\source\repos\runtime-fork\artifacts\bin\dotnet-latest\packs\Microsoft.NET.Runtime.WebAssembly.Wasi.Sdk\9.0.0-dev\Sdk\WasiApp.Native.targets(273,5): error : C:\Users\source\repos\runtime-fork\artifacts\bin\Wasi.Build.Tests\Release\net8.0\wasi-wasm\wbt\Debug_dslxjqc1_otc>set errorlevel=dummy  [C:\Users\source\repos\runtime-fork\artifacts\bin\Wasi.Build.Tests\Release\net8.0\wasi-wasm\wbt\Debug_dslxjqc1_otc\Debug_dslxjqc1_otc.csproj]
          [Debug_dslxjqc1_otc] C:\Users\source\repos\runtime-fork\artifacts\bin\dotnet-latest\packs\Microsoft.NET.Runtime.WebAssembly.Wasi.Sdk\9.0.0-dev\Sdk\WasiApp.Native.targets(273,5): error : C:\Users\source\repos\runtime-fork\artifacts\bin\Wasi.Build.Tests\Release\net8.0\wasi-wasm\wbt\Debug_dslxjqc1_otc>set errorlevel=  [C:\Users\source\repos\runtime-fork\artifacts\bin\Wasi.Build.Tests\Release\net8.0\wasi-wasm\wbt\Debug_dslxjqc1_otc\Debug_dslxjqc1_otc.csproj]
          [Debug_dslxjqc1_otc] C:\Users\source\repos\runtime-fork\artifacts\bin\dotnet-latest\packs\Microsoft.NET.Runtime.WebAssembly.Wasi.Sdk\9.0.0-dev\Sdk\WasiApp.Native.targets(273,5): error : C:\Users\source\repos\runtime-fork\artifacts\bin\Wasi.Build.Tests\Release\net8.0\wasi-wasm\wbt\Debug_dslxjqc1_otc>C:\Users\source\repos\runtime-fork\src\mono\wasi\wasi-sdk\bin\clang.exe "@C:\Users\source\repos\runtime-fork\artifacts\bin\dotnet-latest\packs\Microsoft.NETCore.App.Runtime.Mono.wasi-wasm\9.0.0-dev\runtimes\wasi-wasm\native\src\wasi-default.rsp"  "@C:\Users\source\repos\runtime-fork\artifacts\bin\Wasi.Build.Tests\Release\net8.0\wasi-wasm\wbt\Debug_dslxjqc1_otc\obj\Debug\net8.0\wasi-wasm\wasm\for-build\wasi-compile.rsp" -c -o "C:\Users\AppData\Local\Temp\tmpiipfiv.tmp" "C:\Users\source\repos\runtime-fork\artifacts\bin\dotnet-latest\packs\Microsoft.NETCore.App.Runtime.Mono.wasi-wasm\9.0.0-dev\runtimes\wasi-wasm\native\src\driver.c"  [C:\Users\source\repos\runtime-fork\artifacts\bin\Wasi.Build.Tests\Release\net8.0\wasi-wasm\wbt\Debug_dslxjqc1_otc\Debug_dslxjqc1_otc.csproj]
          [Debug_dslxjqc1_otc] C:\Users\source\repos\runtime-fork\artifacts\bin\dotnet-latest\packs\Microsoft.NET.Runtime.WebAssembly.Wasi.Sdk\9.0.0-dev\Sdk\WasiApp.Native.targets(273,5): error : C:\Users\source\repos\runtime-fork\artifacts\bin\dotnet-latest\packs\Microsoft.NETCore.App.Runtime.Mono.wasi-wasm\9.0.0-dev\runtimes\wasi-wasm\native\src\driver.c:473:2: error: call to undeclared function 'register_aot_modules'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration] [C:\Users\source\repos\runtime-fork\artifacts\bin\Wasi.Build.Tests\Release\net8.0\wasi-wasm\wbt\Debug_dslxjqc1_otc\Debug_dslxjqc1_otc.csproj]
          [Debug_dslxjqc1_otc] C:\Users\source\repos\runtime-fork\artifacts\bin\dotnet-latest\packs\Microsoft.NET.Runtime.WebAssembly.Wasi.Sdk\9.0.0-dev\Sdk\WasiApp.Native.targets(273,5): error :         register_aot_modules (); [C:\Users\source\repos\runtime-fork\artifacts\bin\Wasi.Build.Tests\Release\net8.0\wasi-wasm\wbt\Debug_dslxjqc1_otc\Debug_dslxjqc1_otc.csproj]
          [Debug_dslxjqc1_otc] C:\Users\source\repos\runtime-fork\artifacts\bin\dotnet-latest\packs\Microsoft.NET.Runtime.WebAssembly.Wasi.Sdk\9.0.0-dev\Sdk\WasiApp.Native.targets(273,5): error :         ^ [C:\Users\source\repos\runtime-fork\artifacts\bin\Wasi.Build.Tests\Release\net8.0\wasi-wasm\wbt\Debug_dslxjqc1_otc\Debug_dslxjqc1_otc.csproj]
          [Debug_dslxjqc1_otc] C:\Users\source\repos\runtime-fork\artifacts\bin\dotnet-latest\packs\Microsoft.NET.Runtime.WebAssembly.Wasi.Sdk\9.0.0-dev\Sdk\WasiApp.Native.targets(273,5): error : 1 error generated. [took 0,39s] [C:\Users\source\repos\runtime-fork\artifacts\bin\Wasi.Build.Tests\Release\net8.0\wasi-wasm\wbt\Debug_dslxjqc1_otc\Debug_dslxjqc1_otc.csproj]
          [Debug_dslxjqc1_otc] 
          [Debug_dslxjqc1_otc] Build FAILED.
```

--- a/src/mono/wasi/build/WasiApp.Native.targets
+++ b/src/mono/wasi/build/WasiApp.Native.targets
@@ -156,6 +156,9 @@
     </PropertyGroup>
 
     <ItemGroup>
+      <WasiAfterRuntimeLoadedDeclarations Include="@(WasiAfterRuntimeLoaded->'void %(Identity)();')" />
+      <WasiAfterRuntimeLoadedCalls Include="@(WasiAfterRuntimeLoaded->'%(Identity)();')" />
+
       <_WasmCommonIncludePaths Include="$(_WasmIntermediateOutputPath.TrimEnd('\/'))" />
       <_WasmCommonIncludePaths Include="$(_WasmRuntimePackIncludeDir)mono-2.0" />
       <_WasmCommonIncludePaths Include="$(_WasmRuntimePackIncludeDir)wasm" />
@@ -410,13 +413,7 @@
     <Delete Files="@(WasmBundleFileToDelete)" />
   </Target>
 
-  <!-- replace with wasmlinkdotnet -->
-  <Target Name="_WasiLinkDotNet"
-          DependsOnTargets="_CheckWasiClangIsExpectedVersion;_WasmSelectRuntimeComponentsForLinking;_WasmCompileAssemblyBitCodeFilesForAOT;_WasmWriteRspFilesForLinking"
-          Inputs="@(_WasmLinkDependencies);$(_WasiClangDefaultFlagsRsp);$(_WasiClangDefaultLinkFlagsRsp);$(_WasiClangLinkRsp)"
-          Outputs="$(_WasmOutputFileName)"
-          Returns="@(FileWrites)">
-
+  <Target Name="_WasmWriteRspFilesForLinking" DependsOnTargets="_CheckWasiClangIsExpectedVersion;_WasmCalculateInitialHeapSize">
     <!-- Generate a file entrypoint_YourAssemblyName.c containing the dotnet_wasi_getentrypointassemblyname symbol.
        This means we don't have to hardcode the assembly name in main.c -->
     <PropertyGroup>
@@ -434,9 +431,9 @@
               WriteOnlyWhenDifferent="true" />
     <ItemGroup>
       <FileWrites Include="$(_WasiGetEntrypointCFile)" />
-      <WasiAfterRuntimeLoadedDeclarations Include="@(WasiAfterRuntimeLoaded->'void %(Identity)();')" />
-      <WasiAfterRuntimeLoadedCalls Include="@(WasiAfterRuntimeLoaded->'%(Identity)();')" />
+    </ItemGroup>
 
+    <ItemGroup>
       <_WasiClangXLinkerFlags Include="--initial-memory=$(WasmInitialHeapSize)" />
 
       <_WasmNativeFileForLinking Include="@(NativeFileReference)" />
@@ -444,6 +441,8 @@
       <_WasiFilePathForFixup Include="$(_WasiGetEntrypointCFile)" />
       <_WasiFilePathForFixup Include="@(_WasiObjectFilesForBundle)" />
       <_WasiFilePathForFixup Include="@(_WasmNativeFileForLinking)" />
+
+      <_WasmLinkDependencies Include="@(_WasiFilePathForFixup)" />
 
       <_WasiSdkClangArgs Condition="'$(OS)' == 'Windows_NT'" Include="&quot;$([System.String]::new(%(_WasiFilePathForFixup.Identity)).Replace('\', '/'))&quot;" />
       <_WasiSdkClangArgs Condition="'$(OS)' != 'Windows_NT'" Include="@(_WasiFilePathForFixup -> '&quot;%(Identity)&quot;')" />
@@ -460,7 +459,18 @@
       <_WasiSdkClangArgs Include="-o &quot;$(_WasmOutputFileName.Replace('\', '/'))&quot;" />
     </ItemGroup>
 
-    <WriteLinesToFile Lines="@(_WasiSdkClangArgs)" File="$(_WasiClangLinkRsp)" Overwrite="true" />
+    <WriteLinesToFile Lines="@(_WasiSdkClangArgs)" File="$(_WasiClangLinkRsp)" Overwrite="true" WriteOnlyWhenDifferent="true" />
+    <ItemGroup>
+      <_WasmLinkDependencies Include="$(_WasiClangLinkRsp)" />
+    </ItemGroup>
+  </Target>
+
+  <!-- replace with wasmlinkdotnet -->
+  <Target Name="_WasiLinkDotNet"
+          DependsOnTargets="_CheckWasiClangIsExpectedVersion;_WasmSelectRuntimeComponentsForLinking;_WasmCompileAssemblyBitCodeFilesForAOT;_WasmWriteRspFilesForLinking"
+          Inputs="@(_WasmLinkDependencies);$(_WasiClangDefaultFlagsRsp);$(_WasiClangDefaultLinkFlagsRsp);$(_WasiClangLinkRsp)"
+          Outputs="$(_WasmOutputFileName)"
+          Returns="@(FileWrites)">
 
     <!--<Message Importance="High" Text="Performing WASI SDK build: &quot;$(WasiClang)&quot; @(_WasiSdkClangArgs, ' ')" />-->
     <Message Importance="High" Text="Performing WASI SDK build: &quot;$(WasiClang)&quot; &quot;$(_WasiClangLinkRsp)&quot;" />
@@ -614,5 +624,5 @@
   </Target>
 
   <Target Name="_CheckWasiClangIsExpectedVersion" />
-  <Target Name="_WasmWriteRspFilesForLinking" />
+  <Target Name="_WasmCalculateInitialHeapSize" />
 </Project>

